### PR TITLE
Permissions first implementation

### PR DESCRIPTION
**Pull request by andreievg**
Fixes: #108, #109

### Brief PR description

This is the initial permissions 'skeleton' PR. Added schema changes, JWT token generation, row level policy generation and added further docs.

#### 02_jwtConfig.sql

Refined to have only the following methods:

* `jwt_get_text`
* `jwt_get_boolean`
* `jwt_get_bigint`

All of the methods check 'session' variable that graphile sets from JWT

#### 13_template_permission.sql

* Added `all_permissions` view to amalgamate all permisison tables

#### 35_row_level_policy_related.sql

* Added a role with login right 'graphile_user' with privileges on all tables
* Enabled row level security for application table

graphile_user can be used in the command line that launches graphile

application table will be restricted by row level policies unlesss user accessing db is postgres

#### insertData.js

* Added rules to permission policies 
   * `oneTimeApply`
   * `basicApply`
   * `basicReview`
* Added a call to 'updateRowPolicies' end point so that row policies will be generated an applied

### API.md

* Added login and userInfo end point description
* Added updateRowPolicies end point description

### package.json

* Added `pg_permissions` script, to launch graphile as `graphile_user`
* Added `json-sql-builder2` dependency

### database helpers

* Changed `getUserTemplatePermissions` to use all_permissions view
* Added `getAllPermission` just calls all_permissions
* Added `getAllGeneratedRowPolicies` which shows all generated policies (from postgres schema table `pg_policies`)

### logic

Reshuffled permission.ts, made a folder with following structure

Screen Shot 2021-01-11 at 12 00 23 PM

#### routes.ts

Just has the three routes, and imports from `loginHelpers` and `rowLevelPolicyHelpers`

#### loginHelpers.ts

Should be unchanged, code extracted from previous permission.ts

#### rowLevelPolicyHelpers

Logic to generate JWT `compileJWT` and to generate and apply row level policies `updateRowPolicies`

#### tests.ts

I tried to extract core logic for both generating JWT and updating row level policies into 'functional' methods that don't interact with DB. These methods can then be easily tested.

They are `compileJWT` and `generateRowLevelPolicies`

#### testdata.ts

Both data sets, test 1 = data from method header comments, test 2 = data from DB

## How to see it run/work ?

If you `database_init` with both graphile (as postgres user) and server running, row level policies should be enabled and installed for application table

You would at this stage need to restart graphile with different user `pg_permissions` script.

### Check with graphile UI

* Try to query for applications, you should see zero records
* With postman get JWT from login endpoint
* Use JWT in headers for web graphile UI (: "Bearer ${JWT}" )
* You should see list of application by application types available for that user

### Check with front end

Use the following branch: https://github.com/openmsupply/application-manager-web-app/pull/257, and see different list of application types based on logged in user

 ## NOTE

 Create, update and delete policies are not configured in insertData for application, thus you would need to start graphile with normal `pg` command to create/update applications

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -16,6 +16,9 @@ To run in dev mode:
 To launch Postgraphile server (in a new Terminal process):  
 `yarn pg`
 
+To launch Postgraphile server as graphile_user (in a new Terminal process), this will enforce row level permissions:  
+`yarn pg_permissions`
+
 To launch in dev mode _with_ Postgraphile server:  
 `yarn dev_pg`
```
Comment (CarlosNZ) [inline README.md:24]:
```suggestion

To launch in dev mode _with_ Postgraphile server with row-level permissions:  
`yarn dev_pg_permissions`
```
See my later comment for adding this additional script (in package.json)

File: src/components/permissions/rowLevelPolicyHelpers.ts [hunk 2]
```diff
@@ -0,0 +1,327 @@
...
+      [permissionAbbreviation]: true,
+      ...remapObjectKeysWithPrefix(permissionAbbreviation, {
+        ...templatePermissionRestrictions,
+        templateId,
+      }),
+      ...JWT,
+    }
+  })
+  return JWT
+}
```
Comment (CarlosNZ) [inline src/components/permissions/rowLevelPolicyHelpers.ts:78]: Just looking at the returned JWT, the username is not present, just the userId. Don't we need the username as well?

**Review by CarlosNZ — COMMENTED** [review]
Just a couple of quick comments and questions. Looks great.

I'll have to do a bit of re-jigging my passwordHashing PR to work in with this, so let's get this one merged first and then I'll fix mine.

**Review by nmadruga — CHANGES_REQUESTED** [review]
Partial review. Still need to have another look on `rowLevelPolicyHelpers` and the tests. A lot to take : ) But so far I have made a few comments and suggestions to be addressed please.

Thank you!

File: documentation/API.md [hunk 3]
```diff
@@ -74,3 +74,99 @@ Request will return an object, structured like so:
...
+- User Info
+- User Permissions
+- JWT token
+
+##### REQUEST Body:
+
+```JSON
+{
+    "username": "${username}"
+    "passwordHash": "${passwordHash"
```
[the remark this answers is not in the document]
Comment (CarlosNZ) [reply documentation/API.md:93]: @nmadruga Don't worry, I'm currently updating my PR to integrate with this, so I'll update the docs in there too.

File: src/components/permissions/routes.ts [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+import { updateRowPolicies } from './rowLevelPolicyHelpers'
+
+const routeUserPermissions = async (request: any, reply: any) => {
+  const token = (request?.headers?.authorization || '').replace('Bearer ', '')
+  const username = await getUsername(token)
+  return reply.send(await getUserInfo(username))
+}
+
+const routeLogin = async (request: any, reply: any) => {
+  const { username, passwordHash } = request.body || { username: '', passwordHash: '' }
```
[the remark this answers is not in the document]
Comment (CarlosNZ) [reply src/components/permissions/routes.ts:12]: I'm doing it :) (in my PR)

File: src/components/permissions/loginHelpers.ts [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+import { compileJWT } from './rowLevelPolicyHelpers'
+
+const verifyPromise: any = promisify(verify)
+const signPromise: any = promisify(sign)
+
+const getUsername = async (jwtToken: string) => {
+  let username = 'nonRegistered'
+  if (jwtToken) {
+    try {
+      username = (await verifyPromise(jwtToken, config.jwtSecret)).user.username
```
Comment (CarlosNZ) [inline src/components/permissions/loginHelpers.ts:15]:
```suggestion
      username = (await verifyPromise(jwtToken, config.jwtSecret)).username
```
Think it should be this, username is root level of JWT. (it will be once you re-instate it ;)

File: package.json [hunk 6]
```diff
@@ -34,6 +34,7 @@
     "build": "rimraf ./build && tsc && rsync -a --exclude \"*/node_modules\" ./src/plugins ./build",
     "test": "jest",
     "pg": "exec ./utils/postgraphile_launch.sh",
+    "pg_permissions": "exec ./utils/postgraphile_launch_permissions.sh",
     "dev_pg": "concurrently --kill-others-on-fail \"yarn dev\"  \"yarn pg\"",
```
[the remark this answers is not in the document]
Comment (andreievg) [reply package.json:38]: 0ef20cc3de12ee877e283e4c71a77deebfa0420e

File: README.md (hunk 1 above)
Comment (andreievg) [reply README.md:28]: 0ef20cc3de12ee877e283e4c71a77deebfa0420e

File: src/components/permissions/rowLevelPolicyHelpers.ts (hunk 2 above)
Comment (andreievg) [reply src/components/permissions/rowLevelPolicyHelpers.ts]: 12b4c98e39d006c54e02522f8e6c5189497d459a

Also I made an issue to address this properly: https://github.com/openmsupply/application-manager-server/issues/178

File: database/insertData.js [hunk 7]
```diff
@@ -2080,52 +2094,65 @@ const queries = [
...
+                    }
+                    permissionPolicyToPermissionPolicyId: {
+                      create: {
+                        type: REVIEW
+                        name: "basicReview"
+                        rules: {
+                          application: {
+                            view: {
+                              template_id: "jwtPermission_bigint_templateId"
+                            }
```
[the remark this answers is not in the document]
Comment (andreievg) [reply database/insertData.js:2122]: Yeah this is very basic example, we would need to add a lot of configurations for permissions to work correctly.

Would need a process to manage it (i.e. to make sure all tables are correctly restricted when time comes to deploy).

File: database/insertData.js [hunk 8]
```diff
@@ -2080,52 +2094,65 @@ const queries = [
...
+              { permissionNameId: 1 }
+              { permissionNameId: 2 }
+              {
+                permissionNameToPermissionNameId: {
+                  create: {
+                    name: "reviewCompanyRego"
+                    templatePermissionsUsingId: {
+                      create: [{ templateId: 2, restrictions: { stage: 1 } }]
+                    }
+                    permissionPolicyToPermissionPolicyId: {
```
[the remark this answers is not in the document]
Comment (andreievg) [reply database/insertData.js:2114]: I think so, this is form graphql UI:
Screen Shot 2021-01-13 at 12 43 51 PM

You can see correct field names in permission_policy -> create

File: documentation/API.md [hunk 9]
```diff
@@ -74,3 +74,99 @@ Request will return an object, structured like so:
...
+            "Apply",
+            "Review"
+        ]
+    },
+    "JWT": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJwcDNwbjN0cDMiOnRydWUsInBwM3BuM3RwM19zdGFnZSI6IjEiLCJwcDNwbjN0cDNfdGVtcGxhdGVJZCI6IjIiLCJwcDNwbjMiOnRydWUsInBwMnBuMnRwMiI6dHJ1ZSwicHAycG4ydHAyX3RlbXBsYXRlSWQiOiIyIiwicHAycG4yIjp0cnVlLCJwcDFwbjF0cDEiOnRydWUsInBwMXBuMXRwMV90ZW1wbGF0ZUlkIjoiMSIsInBwMXBuMSI6dHJ1ZSwidXNlcklkIjoxMCwiYXVkIjoicG9zdGdyYXBoaWxlIiwiaWF0IjoxNjEwMDYwMTQxfQ.X8E7CpwRfVNHgm5wW6Qxp122vZmLr8cwipPXFcx_YUo",
+    "user": {
+        "id": 10,
+        "firstName": null,
+        "lastName": null,
+        "userId": 10,
```
[the remark this answers is not in the document]
Comment (andreievg) [reply documentation/API.md:116]: Sure, this is not quite from this PR, but I can change it in this PR (just not sure if both fields are used somewhere in front end ?)

File: documentation/API.md [hunk 10]
```diff
@@ -74,3 +74,99 @@ Request will return an object, structured like so:
...
+    "CREATE POLICY \"view_pp3pn3\" ON \"application\" FOR SELECT USING (jwt_get_boolean('pp3pn3') = true and template_id = jwt_get_bigint('pp3pn3_templateId')) ",
+    "CREATE POLICY \"view_pp3pn3tp3\" ON \"application\" FOR SELECT USING (jwt_get_boolean('pp3pn3tp3') = true and template_id = jwt_get_bigint('pp3pn3tp3_templateId')) "
+]
+```
+
+**To-do**: authorisation of some sort (use JWT, and some way to define ADMIN user, maybe boolean on permission_policy)
+
+**To-do**: fine grained change vs dropping all and reinstating them
+
+**To-do**: add link to permission wiki
```
[the remark this answers is not in the document]
Comment (andreievg) [reply documentation/API.md]: I think so: d266b8f7f05cb2e2c033d3b5ccdee81fb7451bbf

File: src/components/permissions/loginHelpers.ts [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+
+  return {
+    templatePermissions: buildTemplatePermissions(templatePermissionRows),
+    JWT: await getSignedJWT(userInfo, templatePermissionRows),
+    user: userInfo,
+  }
+}
+
+const buildTemplatePermissions = (templatePermissionRows: Array<PermissionRow>) => {
+  const templatePermissions: TemplatePermissions = {} // TODO add type
```
[the remark this answers is not in the document]
Comment (andreievg) [reply src/components/permissions/loginHelpers.ts]: 491b6dc8ecfc06bd6222faf3acd1b438876d05af

File: src/components/postgresConnect.ts [hunk 12]
```diff
@@ -381,6 +381,7 @@ class PostgresDB {
       SELECT id,
       first_name as "firstName",
       last_name as "lastName",
+      id as "userId",
```
[the remark this answers is not in the document]
Comment (andreievg) [reply src/components/postgresConnect.ts:384]: Hmm maybe i should change my PR to just use the id, was wondering, what does front end use ?

File: src/components/permissions/rowLevelPolicyHelpers.ts [hunk 13]
```diff
@@ -0,0 +1,327 @@
...
+    JWT = {
+      [permissionAbbreviation]: true,
+      ...remapObjectKeysWithPrefix(permissionAbbreviation, {
+        ...templatePermissionRestrictions,
+        templateId,
+      }),
+      ...JWT,
+    }
+  })
+  return JWT
```
[the remark this answers is not in the document]
Comment (andreievg) [reply src/components/permissions/rowLevelPolicyHelpers.ts]: Hmm, personally it's more readable for me in the for each format, what do you think ?
https://github.com/openmsupply/application-manager-server/blob/78363a9416d564bd2f78800b62bccf07e0ac8cc1/src/components/permissions/rowLevelPolicyHelpers.ts#L52-L77

[the remark this answers is not in the document]
Comment (nmadruga) [reply src/components/permissions/rowLevelPolicyHelpers.ts]: I liked more with the reduce, Idk. It's just more straight forward to me. Althought this bit inside is a little tricky to understand. I'm getting confused with the reuse of `permissionAbbreviation`. Maybe like this would be clearer:

```
const permissionAbbreviation = getPermissionNameAbbreviation(permissionRow) 

if (! templatePermissionId) 
  return { 
         [permissionAbbreviation]: true, 
         ...JWT, 
       }
else {
const templatePermissionAbbreviation = getTemplatePermissionAbbreviation(permissionRow)

   return {
         [permissionAbbreviation]: true,
         [templatePermissionAbbreviation]: true, 
         ...remapObjectKeysWithPrefix(permissionAbbreviation, { 
           ...templatePermissionRestrictions, 
           templateId, 
         }), 
         ...JWT, 
       } 
}
```

Is this the same as the above? I haven't tested this code.

File: src/components/postgresConnect.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (nmadruga) [reply src/components/postgresConnect.ts:384]: Yes, if that's not going to change this code a lot. Otherwise we can change on the frond end no problem. For now it uses `id`.

File: src/components/permissions/routes.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (nmadruga) [reply src/components/permissions/routes.ts:12]: Cool

File: documentation/API.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (nmadruga) [reply documentation/API.md:93]: Yep. No problem then

File: database/insertData.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (nmadruga) [reply database/insertData.js:2122]: Thanks for the explanation and addition. Yes, having something to check all that will be great - can you add a briefing to  some issue (about the process to manage this) so we don't loose this idea?

File: database/insertData.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (nmadruga) [reply database/insertData.js:2114]: Cool thanks for checking! I was suspicious because it's a relation from permissionName to permissionPolicy. So this is all good : )

**Comment by nmadruga** [comment]
After running the tests I got few errors:

1
Screen Shot 2021-01-13 at 4 17 21 PM
2
Screen Shot 2021-01-13 at 4 17 55 PM

3
Screen Shot 2021-01-13 at 4 13 57 PM

**Comment by andreievg** [comment]
@nmadruga `After running the tests I got few errors:`

I've re ran tests, there were some failures (different to yours, which I fixed in): c14c72bf486dc15c222a4771cbf725dff9f5300b

after that I get this:

Screen Shot 2021-01-13 at 4 36 57 PM

For test to run correctly, we need to do a fresh `database_init` with both `yarn pg` and `yarn dev` running, then `yarn test` should work correctly

File: src/components/permissions/rowLevelPolicyHelpers.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (andreievg) [reply src/components/permissions/rowLevelPolicyHelpers.ts]: > Is this the same as the above? I haven't tested this code.

Pretty much `remapObjectKeysWithPrefix(permissionAbbreviation, { ` would need to have templatePermissionAbbreviation rather then permissionAbbreviation.

> I'm getting confused with the reuse of permissionAbbreviation

Interesting enough i had it like you have it previously (still in forEach).

Anyways, I am not convinced that reduce is more readable. If you feel strongly about please ask Carl to be the tie breaker. (sorry I would usually just make the requested change, but reduce feels a bit weird to me here, especially since the initial `{ userId, username, aud: 'postgraphile' } ` seems kind of hidden)

I've made more comments and removed reuse of permissionAbbreviation though.

c6876efc6868b292b066fa68490587a4d3f11143

**Comment by CarlosNZ** [comment]
Would you mind merging the latest `master` into this branch? This branch is currently quite a bit behind.

**Comment by andreievg** [comment]
> @andreievg Would you mind merging the latest master into this branch? This branch is currently quite a bit behind.

Done

File: src/components/postgresConnect.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (andreievg) [reply src/components/postgresConnect.ts:384]: I think Carl is going to attempt the change as part of the login company PR

[the remark this answers is not in the document]
Comment (CarlosNZ) [reply src/components/postgresConnect.ts:384]: Yeah, I've done it -- just kept `userId` and not `id`

File: src/components/permissions/rowLevelPolicyHelpers.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (nmadruga) [reply src/components/permissions/rowLevelPolicyHelpers.ts]: The change you had previously in forEach could have been kept in the `reduce` version to keep it simpler too. 
I still think it would look better, but I really don't think it's worth a big argument around this. Maybe is just a matter of preference.

I'm going to have another look on `rowLevelPolicyHelpers.ts ` I couldn't see clearly what the code is doing 2 steps ahead. That's why I wanted to help with small changes to see if it would be clearer. I always think that it could look simpler even though it's doing hardcore stuff.

**Comment by nmadruga** [comment]
> I've re ran tests, there were some failures (different to yours, which I fixed in): [c14c72b](https://github.com/openmsupply/application-manager-server/commit/c14c72bf486dc15c222a4771cbf725dff9f5300b)
> 
> after that I get this:
> 
> Screen Shot 2021-01-13 at 4 36 57 PM
> 
> For test to run correctly, we need to do a fresh `database_init` with both `yarn pg` and `yarn dev` running, then `yarn test` should work correctly

I'll try again. I think I was using `pg_permissions` which may be why it didn't run all tests. I'll try again now with `pg` instead

File: src/components/permissions/rowLevelPolicyHelpers.ts [hunk 14]
```diff
@@ -0,0 +1,332 @@
...
+// out { prefixMe_one: "1", prefixMe_two: "2" }
+const remapObjectKeysWithPrefix = (prefix: string, object: Object) => {
+  const remappedObject: { [index: string]: string } = {}
+
+  Object.entries(object).forEach(([key, value]) => {
+    remappedObject[`${prefix}_${key}`] = String(value)
+  })
+
+  return remappedObject
+}
```
Comment (nmadruga) [inline src/components/permissions/rowLevelPolicyHelpers.ts:28]: Not a big deal but perhaps could move this function helpers to separated file. To make this one more focused on the final goal => to generate the row level permissions. Whereas the other functions look more like regular utility functions.

File: src/components/permissions/rowLevelPolicyHelpers.ts [hunk 15]
```diff
@@ -0,0 +1,332 @@
...
+        ...templatePermissionRestrictions,
+        templateId,
+      }),
+      ...JWT,
+    }
+  })
+  return JWT
+}
+
+// Removes previously generated row level policies and resintates them based on current permission settings
```
Comment (nmadruga) [inline src/components/permissions/rowLevelPolicyHelpers.ts:85]:
```suggestion
// Removes previously generated row level policies and reinstates them based on current permission settings
```

File: src/components/permissions/rowLevelPolicyHelpers.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (andreievg) [reply src/components/permissions/rowLevelPolicyHelpers.ts]: Cools, I was hoping we could reach a compromise with comments made in this commit https://github.com/openmsupply/application-manager-server/commit/c6876efc6868b292b066fa68490587a4d3f11143

**Comment by andreievg** [comment]
@nmadruga 
>I'll try again. I think I was using pg_permissions which may be why it didn't run all tests. I'll try again now with pg instead.

Just added clarification to README.md 40e8177443a7de503007986d471cb06b7e4c236c

File: src/components/permissions/rowLevelPolicyHelpers.ts [hunk 16]
```diff
@@ -0,0 +1,332 @@
...
+        compileRowLevelPolicy(
+          tableName,
+          permissionAbbreviation,
+          ruleType,
+          usingCondition,
+          condition
+        )
+      )
+    })
+  })
```
Comment (nmadruga) [inline src/components/permissions/rowLevelPolicyHelpers.ts:206]: My personal preference is for `map` instead of `forEach` here

File: src/components/permissions/rowLevelPolicyHelpers.ts [hunk 17]
```diff
@@ -0,0 +1,332 @@
...
+    for: 'CREATE',
+    using: false,
+    withCheck: true,
+  },
+  delete: {
+    for: 'DELETE',
+    using: true,
+    withCheck: false,
+  },
+}
```
Comment (nmadruga) [inline src/components/permissions/rowLevelPolicyHelpers.ts:234]: In separated file file would be nice

File: src/components/permissions/rowLevelPolicyHelpers.ts [hunk 18]
```diff
@@ -0,0 +1,332 @@
...
+  permissionAbbreviation: string,
+  ruleType: string,
+  usingCondition: string,
+  condition: string
+) => {
+  // We want to first check if the permission exists in JWT token (to make sure we only do the full query if it exists),
+  // so we add a check for "policy": true, i.e. jwt_get_boolean('pp2pn2tp2') = true
+  // this will become first condition in both USING and WITH CHECK of every row level rule
+  const addBracketsAndPermissionCheck = (condition: string) =>
+    `(jwt_get_boolean('${permissionAbbreviation}') = true and ${condition})`
```
Comment (nmadruga) [inline src/components/permissions/rowLevelPolicyHelpers.ts:252]: This is very well explained here. Thanks!

File: src/components/permissions/rowLevelPolicyHelpers.ts [hunk 19]
```diff
@@ -0,0 +1,332 @@
...
+    let match
+    const regex = new RegExp(`${prefix}\\w*`, 'g')
+    while ((match = regex.exec(resultSql))) {
+      const matchedValue = match[0]
+      const newValue = `${matchedValue.replace(prefix, prefixReplacement)}${postfix}`
+      resultSql = resultSql.replace(matchedValue, newValue)
+    }
+  })
+  return resultSql
+}
```
Comment (nmadruga) [inline src/components/permissions/rowLevelPolicyHelpers.ts:330]: This one also looks like it could be in a separated file. Just keep the core stuff in this one. Then it might be easier to follow. I'm still trying and it's quite hard.

File: src/components/permissions/tests/testdata.ts [hunk 20]
```diff
@@ -0,0 +1,312 @@
...
+        permissionType: 'Apply',
+        permissionPolicyId: 1,
+        permissionPolicyRules: {},
+        permissionNameId: 1,
+        templatePermissionId: 1,
+        templatePermissionRestrictions: null,
+        templateId: 1,
+        templateCode: 'TestRego',
+        userId: 10,
+        username: 'userWithMultiplePermissions',
```
Comment (nmadruga) [inline src/components/permissions/tests/testdata.ts:57]: I don't understand why the `username` is here as well

**Review by nmadruga — APPROVED** [review]
Ok. Cool I my suggested changes can be considered trivial. The logic is awesome, it's really cool that this can be build in such a fine way. But it's really hard to see what is going on. 

The comments help, and some times overwhelm who is reading. The `rowLevelPolicyHelpers` seems to be the heart of everything, but since it contains everything is quite hard to follow the line. Not a big deal if no one else will be touching the code, but since we need to agree on reviews and all I think it could be better to read if break-down a little more.

File: src/components/permissions/rowLevelPolicyHelpers.ts [hunk 21]
```diff
@@ -0,0 +1,332 @@
+import databaseConnect from '../databaseConnect'
+import { UserInfo, PermissionRow, RuleTypes } from './types'
+
+const SQLBuilder = require('json-sql-builder2')
```
[the remark this answers is not in the document]
Comment (andreievg) [reply src/components/permissions/rowLevelPolicyHelpers.ts:4]: No typescript definition available, this was the quicker and easier way to do it

File: src/components/permissions/rowLevelPolicyHelpers.ts (hunk 14 above)
Comment (andreievg) [reply src/components/permissions/rowLevelPolicyHelpers.ts:28]: Ok sounds good but struggling to find a good name for the file, would 'helpersUtils' be appropriate ?

File: src/components/permissions/rowLevelPolicyHelpers.ts (hunk 16 above)
Comment (andreievg) [reply src/components/permissions/rowLevelPolicyHelpers.ts:206]: I think the reason I used forEach and push, it to avoid extra syntax for destructuring.
i.e.
```TS
let allPolicies: Array<string> = []
// ...
const subsetPolicies = Object.entries(rulesByType).map(.....)
allPolicies = [...subsetPolicies, ...allPolicies]
```

File: src/components/permissions/tests/testdata.ts (hunk 20 above)
Comment (andreievg) [reply src/components/permissions/tests/testdata.ts:57]: It's an artifact, was previously there and was removed earlier in the PR and put back again for userInfo end point to work, Carl is changing that I think in his latest PR

File: src/components/permissions/rowLevelPolicyHelpers.ts (hunk 19 above)
Comment (andreievg) [reply src/components/permissions/rowLevelPolicyHelpers.ts:330]: Sure can move it, once we agree on a good name for that utility file.

Btw, just trying to understand, you've made a comment

 > but since it contains everything is quite hard to follow the line

Is it the length of the file that's making it hard to follow, or excessive break down of logic into methods, so you need to jump from one to another (or maybe the method naming) ?

If it's the "excessive break down of logic into methods, so you need to jump from one to another", then would putting some of the methods in another file make it easier or more difficult ?

File: src/components/permissions/rowLevelPolicyHelpers.ts (hunk 17 above)
Comment (andreievg) [reply src/components/permissions/rowLevelPolicyHelpers.ts:234]: Can move it to that utility methods file. I thought to keep it right above where it's used, so it's easer to see what `const ruleSettings = ruleTypes[ruleType]` is doing. It's quite difficult to try and find a readability balance (between abstracting logic into methods/variables and using those methods/variables in a place where you can't directly see what those methods/variables do)

File: src/components/permissions/rowLevelPolicyHelpers.ts (hunk 21 above)
[the remark this answers is not in the document]
Comment (nmadruga) [reply src/components/permissions/rowLevelPolicyHelpers.ts:4]: Ok. Cool

File: src/components/permissions/rowLevelPolicyHelpers.ts (hunk 14 above)
Comment (nmadruga) [reply src/components/permissions/rowLevelPolicyHelpers.ts:28]: `helpersUtils` sound great

Comment (andreievg) [reply src/components/permissions/rowLevelPolicyHelpers.ts]: 436c4a86aa4d642c025d1d430b301194b29e9370

File: src/components/permissions/rowLevelPolicyHelpers.ts (hunk 17 above)
Comment (andreievg) [reply src/components/permissions/rowLevelPolicyHelpers.ts]: 60db69257ccc33df3bd930a57d334fe42862b8bf

File: src/components/permissions/rowLevelPolicyHelpers.ts (hunk 19 above)
Comment (andreievg) [reply src/components/permissions/rowLevelPolicyHelpers.ts:265]: I've move getSqlConditionFromJSON to helpersUtils: dfb94899f966a33a7b50c39a337dcc9869732288

But I've kept `replacePlaceholders` here, since I thin it's part of the core

File: src/components/permissions/rowLevelPolicyHelpers.ts (hunk 17 above)
Comment (andreievg) [reply src/components/permissions/rowLevelPolicyHelpers.ts]: Ok, changed it further as per telegram chat:
4ca8a2d7ce673dfc9e7fb0f219fe9d6fcf69e2d8

Comment (andreievg) [reply src/components/permissions/rowLevelPolicyHelpers.ts]: Btw, from my understanding `model` is kind of like the schema, the description of shape of DB. In this case we have a constant. Same same in terms of your comment but dont' want to get confused with the terms

File: src/components/permissions/rowLevelPolicyHelpers.ts (hunk 16 above)
Comment (andreievg) [reply src/components/permissions/rowLevelPolicyHelpers.ts:188]: I think it's another of those stylistic preferences.

In the forEach case, there is no need for concat or restructuring, so my point is, why add it ?

File: src/components/permissions/rowLevelPolicyHelpers.ts (hunk 19 above)
Comment (andreievg) [reply src/components/permissions/rowLevelPolicyHelpers.ts:266]: > I was thinking why I have some difficult on following the code, you have heaps of comments (which is good in a sense, but would be easier if just reading the code was enough to understand what was going on)

Most of the comment describe input and output of methods, which is necessary in this case to understand what the code is trying to do. I personally not a fan of excessive comment, I think this level of complexity warrant more then just variable/method naming to explain functionality ( and it's one of those cases where it's hard to describe functionality in a few words that can be used to name variables/methods )

>  With so many things in the same file makes it hard to see nested and not-nested functions. 

Hmm, I think the suggestion you've made to move things to say utils methods, make it a bit easier 

> The order of how calls are made seems a bit random to me, but I haven't tried to debug to see, maybe it was just my impression.

it's mainly top down for core functional areas, (i.e. method will call within itself a method declared below), but not for helpers).

**Comment by andreievg** [comment]
It looks like i've dealt with all of the comments. I think if there is anything else, we can move it to an issue @nmadruga @CarlosNZ ?

Merging
